content/advent-2018/fmt.md

```diff
@@ -0,0 +1,304 @@
+		key = keyMask
+	}
+	return fmt.Sprintf("Login:%s, ACL:%08b, APIKey: %s", ai.Login, ai.ACL, key)
+}
+```
+
+And now `fmt.Formatter` which gets a
+[`fmt.State`](https://golang.org/pkg/fmt/#State) and a rune for the verb.
+`fmt.State` implements [`io.Writer`](https://golang.org/pkg/io/#Writer),
+enabling you to write directly to it.
+
+To know all the fields available in a struct, you can use the
+[`reflect`](https://golang.org/pkg/reflect/).  package. This will make sure
+your code works even when `AuthInfo` changes.
+
+```go
+var authInfoFields []string
+
+func init() {
+	typ := reflect.TypeOf(AuthInfo{})
+	authInfoFields = make([]string, typ.NumField())
+	for i := 0; i < typ.NumField(); i++ {
+		authInfoFields[i] = typ.Field(i).Name
+	}
+	sort.Strings(authInfoFields) // People are better with sorted data
+}
+```
+
+And now you're ready to implement `fmt.Formatter`
+```go
+// Format implements fmt.Formatter
+func (ai *AuthInfo) Format(state fmt.State, verb rune) {
+	switch verb {
+	case 's', 'q':
+		val := ai.String()
+		if verb == 'q' {
+			val = fmt.Sprintf("%q", val)
+		}
+		fmt.Fprint(state, val)
+	case 'v':
+		if state.Flag('#') {
+			// Emit type before
+			fmt.Fprintf(state, "%T", ai)
+		}
+		fmt.Fprint(state, "{")
+		val := reflect.ValueOf(*ai)
+		for i, name := range authInfoFields {
+			if state.Flag('#') || state.Flag('+') {
+				fmt.Fprintf(state, "%s:", name)
+			}
+			fld := val.FieldByName(name)
+			if name == "APIKey" && fld.Len() > 0 {
+				fmt.Fprint(state, keyMask)
+			} else {
+				fmt.Fprint(state, fld)
+			}
+			if i < len(authInfoFields)-1 {
+				fmt.Fprint(state, " ")
+			}
+		}
+		fmt.Fprint(state, "}")
+	}
+}
+```
+
+Let's try it out:
+
+```go
+ai := &AuthInfo{
+	Login:  "daffy",
+	ACL:    ReadACL | WriteACL,
+	APIKey: "duck season",
+}
+fmt.Println(ai.String())
+fmt.Printf("ai %%s: %s\n", ai)
+fmt.Printf("ai %%q: %q\n", ai)
+fmt.Printf("ai %%v: %v\n", ai)
+fmt.Printf("ai %%+v: %+v\n", ai)
+fmt.Printf("ai %%#v: %#v\n", ai)
+```
+
+which will emit
+```
+Login:daffy, ACL:00000011, APIKey: *****
+ai %s: Login:daffy, ACL:00000011, APIKey: *****
+ai %q: "Login:daffy, ACL:00000011, APIKey: *****"
+ai %v: {3 ***** daffy}
+ai %+v: {ACL:3 APIKey:***** Login:daffy}
+ai %#v: *main.AuthInfo{ACL:3 APIKey:***** Login:daffy}
+```
+
+# Conculsion
+The `fmt` package has many capabilities other than the trivial use. Once you'll
+familiarize yourself with these capabilities, I'm sure you find many
+interesting uses for them.
+
+You can view the code for this post
+[here](https://github.com/gopheracademy/gopheracademy-web/blob/master/content/advent-2018/fmt.go).
+
+# About the Author
+Hi there, I'm Miki, nice to e-meet you ☺. I've been a long time developer and
+have been working with Go for about 8 years now. I write code professionally as
+a consultant and contribute a lot to open source. Apart from that I'm a [book
+author](https://www.amazon.com/Forging-Python-practices-lessons-developing-ebook/dp/B07C1SH5MP),
+an author on [LinkedIn
```

Comment:
is the newline intentional?

Reply:
I have `textwidth` set to 80 in Vim.  Doesn't impact the markdown - if it bother you I'll fix it.